### PR TITLE
Aggregation API and Geohash Aggregation implementation

### PR DESCRIPTION
**Pull request by sinistersnare**
Hi! 

This partly goes towards #1703. After this is in, creating hex-bin aggregations should be easy enough, and that can be closed.

There are a few TODOs left in here, but mostly its done! Just wanted to see what you guys need from me to get this merged in. Also a code-review would be great.

Thanks!

### CODE REVIEW

File: core/geotime/src/main/java/org/locationtech/geowave/core/geotime/store/query/aggregate/GeohashBinningStrategy.java [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+        bit = 0;
+      }
+    }
+    return geohash.toString();
+  }
+
+  @Override
+  public String binEntry(T entry) {
+    final Geometry geometry;
+    // TODO: may be better to subclass this so we dont need to do type buffoonery?
```
Comment (sinistersnare) [inline core/geotime/src/main/java/org/locationtech/geowave/core/geotime/store/query/aggregate/GeohashBinningStrategy.java:109]: I will do this

File: core/geotime/src/main/java/org/locationtech/geowave/core/geotime/store/query/aggregate/GeohashBinningStrategy.java [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+   * @return A geohash precision that, in the worst case, would contain {@code maxBins} different
+   *         hashes.
+   */
+  private static int getDesiredPrecision(long maxBins) {
+    for (int i = 12; i > 0; i--) {
+      if ((double) maxBins >= Math.pow(32, i)) {
+        return i;
+      }
+    }
+    // TODO: should this be 0, AKA 1 bin for all data? (if so make sure to fix tests!)
```
Comment (sinistersnare) [inline core/geotime/src/main/java/org/locationtech/geowave/core/geotime/store/query/aggregate/GeohashBinningStrategy.java:43]: Making this 0 would be the same as not binning, should that be the behavior here?

File: extensions/adapters/vector/src/main/java/org/locationtech/geowave/adapter/vector/query/aggregation/VectorMathAggregation.java [hunk 3]
```diff
@@ -53,38 +70,24 @@ public void setParameters(final FieldNameParam fieldNameParam) {
...
   }
 
   @Override
-  public BigDecimal getResult() {
-    return value;
+  public BigDecimal merge(final BigDecimal result1, final BigDecimal result2) {
+    return agg(result1, result2);
   }
 
+  // TODO: is it OK that I changed these? I fixed the tests.
```
Comment (sinistersnare) [inline extensions/adapters/vector/src/main/java/org/locationtech/geowave/adapter/vector/query/aggregation/VectorMathAggregation.java:82]: I added VarintUtils.read/writeBigDecimal, so I changed it here too. Is that alright?

File: examples/java-api/src/main/java/org/locationtech/geowave/examples/binning/GeohashBinningExample.java [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+import org.locationtech.jts.geom.Geometry;
+import org.opengis.feature.simple.SimpleFeature;
+import org.opengis.feature.simple.SimpleFeatureType;
+import java.math.BigDecimal;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+
+public class GeohashBinningExample {
```
Comment (sinistersnare) [inline examples/java-api/src/main/java/org/locationtech/geowave/examples/binning/GeohashBinningExample.java:31]: Is this example good? Should I add more docs describing each step?

**Comment by sinistersnare** [comment]
I signed the ECA a while ago, how can I get confirmation of that?

File: extensions/adapters/vector/src/main/java/org/locationtech/geowave/adapter/vector/query/aggregation/VectorMathAggregation.java [hunk 5]
```diff
@@ -53,38 +70,24 @@ public void setParameters(final FieldNameParam fieldNameParam) {
...
   @Override
   public byte[] resultToBinary(BigDecimal result) {
-    if (result == null) {
-      return new byte[0];
-    }
-    final byte[] unscaled = result.unscaledValue().toByteArray();
```
Comment (jdgarrett) [inline extensions/adapters/vector/src/main/java/org/locationtech/geowave/adapter/vector/query/aggregation/VectorMathAggregation.java:70]: You can move this logic to the VarintUtils class since it is using Variable length encoding for each part of the BigDecimal.

File: test/src/test/java/org/locationtech/geowave/test/GeoWaveITSuite.java [hunk 6]
```diff
@@ -56,55 +57,57 @@
 
 @RunWith(GeoWaveITSuiteRunner.class)
 @SuiteClasses({
-    GeoWaveVisibilityIT.class,
```
Comment (jdgarrett) [inline test/src/test/java/org/locationtech/geowave/test/GeoWaveITSuite.java:59]: Comment these back in

File: test/src/test/java/org/locationtech/geowave/test/basic/GeoWaveBinningAggregationIT.java [hunk 7]
```diff
@@ -0,0 +1,133 @@
+package org.locationtech.geowave.test.basic;
```
Comment (jdgarrett) [inline test/src/test/java/org/locationtech/geowave/test/basic/GeoWaveBinningAggregationIT.java:1]: Missing copyright header.

**Comment by jdgarrett** [comment]
@rfecher I noticed when reviewing this that most aggregations simply return `new byte[0]` in their serialization functions.  It the expectation that an aggregation that is serialized and then deserialized is equivalent to the original, including the value?

File: core/store/src/main/java/org/locationtech/geowave/core/store/base/BaseQueryOptions.java [hunk 8]
```diff
@@ -252,7 +243,7 @@ public boolean isAllAdapters() {
           }
         }
       }
-      return adapters.toArray(new InternalDataAdapter[adapters.size()]);
```
[the remark this answers is not in the document]
Comment (sinistersnare) [reply core/store/src/main/java/org/locationtech/geowave/core/store/base/BaseQueryOptions.java:255]: IntelliJ warns about this code, and when I researched it, I came up with this https://stackoverflow.com/questions/174093/

It seems like, counter-intuitively, not providing a size is faster.

File: core/store/src/main/java/org/locationtech/geowave/core/store/query/aggregate/BinningAggregation.java [hunk 9]
```diff
@@ -0,0 +1,134 @@
...
+  public BinningAggregation() {
+    this(null, null);
+  }
+
+  public BinningAggregation(
+      Supplier<Aggregation<P, R, T>> baseAggregation,
+      AggregationBinningStrategy<T> binningStrategy) {
+    this.baseAggregation = baseAggregation;
+    this.binningStrategy = binningStrategy;
+    this.result = Maps.newHashMapWithExpectedSize(1024);
```
[the remark this answers is not in the document]
Comment (sinistersnare) [reply core/store/src/main/java/org/locationtech/geowave/core/store/query/aggregate/BinningAggregation.java]: I changed this already. The BinningStrategy now has a 'sizeHint' default method that an impl can override to give a better estimate. The default returns `100` which is a decidedly non-scientific number.

File: core/store/src/main/java/org/locationtech/geowave/core/store/query/aggregate/BinningAggregation.java [hunk 10]
```diff
@@ -0,0 +1,134 @@
...
+
+  @Override
+  public Map<String, R> getResult() {
+    return this.result.entrySet().stream().collect(
+        Collectors.toMap(Map.Entry::getKey, e -> e.getValue().getResult()));
+  }
+
+  @Override
+  public Map<String, R> merge(Map<String, R> result1, Map<String, R> result2) {
+    Aggregation<P, R, T> agg = this.result.values().stream().findAny().orElse(null);
```
[the remark this answers is not in the document]
Comment (sinistersnare) [reply core/store/src/main/java/org/locationtech/geowave/core/store/query/aggregate/BinningAggregation.java:70]: The issue is that we need an Aggregation to know what to do with it, as the inner aggregation has the `agg::merge` function that we need. In the case that the agg hasnt been ran, I dont know what to do, so I just bail out.

File: test/src/test/java/org/locationtech/geowave/test/GeoWaveITSuite.java (hunk 6 above)
Comment (sinistersnare) [reply test/src/test/java/org/locationtech/geowave/test/GeoWaveITSuite.java:59]: already done, forgot to uncomment it.

File: core/geotime/src/test/java/org/locationtech/geowave/core/geotime/store/query/aggregate/AbstractVectorAggregationTest.java [hunk 11]
```diff
@@ -19,7 +18,9 @@
 import org.locationtech.jts.geom.Coordinate;
 import org.locationtech.jts.geom.Geometry;
 import org.opengis.feature.simple.SimpleFeature;
-import com.google.common.collect.Lists;
+import org.opengis.feature.simple.SimpleFeatureType;
+import java.util.Date;
+import java.util.List;
```
Comment (sinistersnare) [inline core/geotime/src/test/java/org/locationtech/geowave/core/geotime/store/query/aggregate/AbstractVectorAggregationTest.java:23]: need to go thru the import reorganizations and change them back, they were fallout from a different formatter.

File: core/index/src/test/java/org/locationtech/geowave/core/index/VarintUtilsTest.java [hunk 12]
```diff
@@ -220,4 +221,20 @@ private void testEncodeDecodeTimeValue(final Date value) {
...
     Assert.assertEquals(value, decoded);
   }
+
+  @Test
+  public void testEncodeDecodeBigDecimal() {
+    testEncodeDecodeBigDecimalValue(new BigDecimal(123));
+    testEncodeDecodeBigDecimalValue(new BigDecimal(-123));
+    testEncodeDecodeBigDecimalValue(new BigDecimal(256));
+    testEncodeDecodeBigDecimalValue(new BigDecimal(2_061_000_009));
+    testEncodeDecodeBigDecimalValue(new BigDecimal(-1_000_000_000));
```
Comment (sinistersnare) [inline core/index/src/test/java/org/locationtech/geowave/core/index/VarintUtilsTest.java:231]: Add a few tests that are out of range of the primitive types.

File: core/store/src/main/java/org/locationtech/geowave/core/store/base/BaseQueryOptions.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (sinistersnare) [reply core/store/src/main/java/org/locationtech/geowave/core/store/base/BaseQueryOptions.java:255]: If you still want a change thats fine, just providing my reasoning.

[the remark this answers is not in the document]
Comment (jdgarrett) [reply core/store/src/main/java/org/locationtech/geowave/core/store/base/BaseQueryOptions.java:255]: That's fine, if it's been shown to be faster I have no issue with it.

File: core/store/src/main/java/org/locationtech/geowave/core/store/query/aggregate/BinningAggregation.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (jdgarrett) [reply core/store/src/main/java/org/locationtech/geowave/core/store/query/aggregate/BinningAggregation.java:83]: I see the issue now, I think a better solution would probably be to just get a new Aggregation from the `baseAggregation` supplier for use with merging.

File: core/store/src/main/java/org/locationtech/geowave/core/store/query/aggregate/BinningAggregation.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (sinistersnare) [reply core/store/src/main/java/org/locationtech/geowave/core/store/query/aggregate/BinningAggregation.java]: I reverted the addition of the `sizeHint` default method., because it can return pretty huge values. Precision 6 geohashes can have around 1Billion values, so the size hint would return that. It caused OOMs in my testing (of course making a 1-billion size hashmap would do that :P)

Any recommendations on what to do? Just using the default `10` of a hashmap probably isnt a good idea.

[the remark this answers is not in the document]
Comment (jdgarrett) [reply core/store/src/main/java/org/locationtech/geowave/core/store/query/aggregate/BinningAggregation.java:76]: Well wouldn't the user be using the `maxBinsAllowed` in the constructor for the GeoHashBinningStrategy?  They would have to have enough memory for 1 billion bins if they specified that many allowed.

File: core/geotime/src/main/java/org/locationtech/geowave/core/geotime/store/query/aggregate/GeohashBinningStrategy.java (hunk 1 above)
Comment (sinistersnare) [reply core/geotime/src/main/java/org/locationtech/geowave/core/geotime/store/query/aggregate/GeohashBinningStrategy.java]: Done this.

File: core/geotime/src/main/java/org/locationtech/geowave/core/geotime/store/query/aggregate/GeohashBinningStrategy.java (hunk 2 above)
Comment (sinistersnare) [reply core/geotime/src/main/java/org/locationtech/geowave/core/geotime/store/query/aggregate/GeohashBinningStrategy.java]: Resolved through on of JD's comments

File: extensions/adapters/vector/src/main/java/org/locationtech/geowave/adapter/vector/query/aggregation/VectorMathAggregation.java (hunk 3 above)
Comment (sinistersnare) [reply extensions/adapters/vector/src/main/java/org/locationtech/geowave/adapter/vector/query/aggregation/VectorMathAggregation.java:82]: Resolved through on of JD's comments

File: examples/java-api/src/main/java/org/locationtech/geowave/examples/binning/GeohashBinningExample.java (hunk 4 above)
Comment (sinistersnare) [reply examples/java-api/src/main/java/org/locationtech/geowave/examples/binning/GeohashBinningExample.java]: Resolved through on of JD's comments

File: core/geotime/src/main/java/org/locationtech/geowave/core/geotime/store/query/aggregate/GeohashBinningStrategy.java [hunk 13]
```diff
@@ -0,0 +1,136 @@
+package org.locationtech.geowave.core.geotime.store.query.aggregate;
```
[the remark this answers is not in the document]
Comment (sinistersnare) [reply core/geotime/src/main/java/org/locationtech/geowave/core/geotime/store/query/aggregate/GeohashBinningStrategy.java:9]: Should be resolved.

File: core/geotime/src/main/java/org/locationtech/geowave/core/geotime/store/query/aggregate/GeohashBinningStrategy.java [hunk 14]
```diff
@@ -0,0 +1,136 @@
...
+   *         hashes.
+   */
+  private static int getDesiredPrecision(long maxBins) {
+    for (int i = 12; i > 0; i--) {
+      if ((double) maxBins >= Math.pow(32, i)) {
+        return i;
+      }
+    }
+    // TODO: should this be 0, AKA 1 bin for all data? (if so make sure to fix tests!)
+    return 1;
```
[the remark this answers is not in the document]
Comment (sinistersnare) [reply core/geotime/src/main/java/org/locationtech/geowave/core/geotime/store/query/aggregate/GeohashBinningStrategy.java:61]: Should be resolved, I implemented this.

File: core/geotime/src/main/java/org/locationtech/geowave/core/geotime/store/query/aggregate/GeohashBinningStrategy.java [hunk 15]
```diff
@@ -0,0 +1,136 @@
...
+import org.opengis.feature.simple.SimpleFeature;
+
+public class GeohashBinningStrategy<T> implements AggregationBinningStrategy<T> {
+
+  private int precision;
+
+  /**
+   * Create a Binning Strategy based on GeoHashes.
+   *
+   * @param maxBinsAllowed TODO: explain how this invariant works, and how itll break for < 32.
```
[the remark this answers is not in the document]
Comment (sinistersnare) [reply core/geotime/src/main/java/org/locationtech/geowave/core/geotime/store/query/aggregate/GeohashBinningStrategy.java]: Should be resolved.

File: core/geotime/src/test/java/org/locationtech/geowave/core/geotime/store/query/aggregate/GeohashBinningStrategyTest.java [hunk 16]
```diff
@@ -0,0 +1,149 @@
+package org.locationtech.geowave.core.geotime.store.query.aggregate;
```
[the remark this answers is not in the document]
Comment (sinistersnare) [reply core/geotime/src/test/java/org/locationtech/geowave/core/geotime/store/query/aggregate/GeohashBinningStrategyTest.java:9]: Should be resolved.

File: core/index/src/main/java/org/locationtech/geowave/core/index/VarintUtils.java [hunk 17]
```diff
@@ -284,4 +272,41 @@ public static void writeTime(final long time, final ByteBuffer buffer) {
...
   }
+
+  /**
+   * Encode a BigDecimal as a byte[]. The structure of the byte[] is opaque, so to deserialize, use
+   * {@link #readBigDecimal(ByteBuffer)}
+   *
+   * @param num The number to serialize as a {@link ByteBuffer}
+   * @return a byte array that represents the given BigDecimal.
+   */
+  public static byte[] writeBigDecimal(final BigDecimal num) {
```
[the remark this answers is not in the document]
Comment (sinistersnare) [reply core/index/src/main/java/org/locationtech/geowave/core/index/VarintUtils.java:283]: Good idea, I just lifted the code I removed from the math-agg and used that instead of this stuff. However, I changed it slightly so it was a better ByteBuffer citizen (storing the array length, instead of using all of `.remaining()`

File: core/store/src/main/java/org/locationtech/geowave/core/store/api/AggregationQueryBuilder.java [hunk 18]
```diff
@@ -43,11 +54,16 @@
   A count(String... typeNames);
 
   /**
-   * get a new default implementation of the builder
+   * Instead of having a scalar aggregation, bin the results by hash.
```
[the remark this answers is not in the document]
Comment (sinistersnare) [reply core/store/src/main/java/org/locationtech/geowave/core/store/api/AggregationQueryBuilder.java]: Good point! Thanks

File: core/store/src/main/java/org/locationtech/geowave/core/store/query/aggregate/AggregationBinningStrategy.java [hunk 19]
```diff
@@ -0,0 +1,7 @@
+package org.locationtech.geowave.core.store.query.aggregate;
```
[the remark this answers is not in the document]
Comment (sinistersnare) [reply core/store/src/main/java/org/locationtech/geowave/core/store/query/aggregate/AggregationBinningStrategy.java:9]: Should be resolved.

File: core/store/src/main/java/org/locationtech/geowave/core/store/query/aggregate/AggregationBinningStrategy.java [hunk 20]
```diff
@@ -0,0 +1,7 @@
...
+
+public interface AggregationBinningStrategy<T> {
+  String binEntry(T entry);
```
[the remark this answers is not in the document]
Comment (sinistersnare) [reply core/store/src/main/java/org/locationtech/geowave/core/store/query/aggregate/AggregationBinningStrategy.java]: Good idea, implemented.

File: core/store/src/main/java/org/locationtech/geowave/core/store/query/aggregate/BinningAggregation.java [hunk 21]
```diff
@@ -0,0 +1,134 @@
+package org.locationtech.geowave.core.store.query.aggregate;
```
[the remark this answers is not in the document]
Comment (sinistersnare) [reply core/store/src/main/java/org/locationtech/geowave/core/store/query/aggregate/BinningAggregation.java:9]: Should be resolved.

File: core/store/src/main/java/org/locationtech/geowave/core/store/query/aggregate/BinningAggregation.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (sinistersnare) [reply core/store/src/main/java/org/locationtech/geowave/core/store/query/aggregate/BinningAggregation.java]: I tried a solution on this, having a 'helper' aggregation that is only used for its functions, not for aggregation.

File: core/store/src/main/java/org/locationtech/geowave/core/store/query/aggregate/BinningAggregation.java [hunk 22]
```diff
@@ -0,0 +1,134 @@
...
+      return null;
+    }
+
+    return Stream.of(result1, result2).flatMap(m -> m.entrySet().stream()).collect(
+        Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue, agg::merge));
+  }
+
+  @Override
+  public byte[] resultToBinary(Map<String, R> result) {
+    Aggregation<P, R, T> agg = this.result.values().stream().findAny().orElse(null);
```
[the remark this answers is not in the document]
Comment (sinistersnare) [reply core/store/src/main/java/org/locationtech/geowave/core/store/query/aggregate/BinningAggregation.java]: This is related to the above conversation, so if that gets resolved, so should this.

File: core/store/src/main/java/org/locationtech/geowave/core/store/query/aggregate/BinningAggregation.java [hunk 23]
```diff
@@ -0,0 +1,134 @@
...
+    ByteBuffer bb = ByteBuffer.allocate(4 + totalDataSize);
+    bb.putInt(mapped.size());
+    mapped.forEach(
+        (k, v) -> bb.putInt(k.length()).put(StringUtils.stringToBinary(k)).putInt(v.length).put(v));
+    return bb.array();
+  }
+
+  @Override
+  public Map<String, R> resultFromBinary(byte[] binary) {
+    Aggregation<P, R, T> agg = this.result.values().stream().findAny().orElse(null);
```
[the remark this answers is not in the document]
Comment (sinistersnare) [reply core/store/src/main/java/org/locationtech/geowave/core/store/query/aggregate/BinningAggregation.java]: This is related to the above conversation, so if that gets resolved, so should this.

File: core/store/src/test/java/org/locationtech/geowave/core/store/query/aggregate/BinningAggregationTest.java [hunk 24]
```diff
@@ -0,0 +1,71 @@
+package org.locationtech.geowave.core.store.query.aggregate;
```
[the remark this answers is not in the document]
Comment (sinistersnare) [reply core/store/src/test/java/org/locationtech/geowave/core/store/query/aggregate/BinningAggregationTest.java:9]: Should be resolved.

File: examples/java-api/src/main/java/org/locationtech/geowave/examples/binning/GeohashBinningExample.java [hunk 25]
```diff
@@ -0,0 +1,123 @@
+package org.locationtech.geowave.examples.binning;
+
```
[the remark this answers is not in the document]
Comment (sinistersnare) [reply examples/java-api/src/main/java/org/locationtech/geowave/examples/binning/GeohashBinningExample.java]: Should be resolved.

File: examples/java-api/src/main/java/org/locationtech/geowave/examples/binning/GeohashBinningExample.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (sinistersnare) [reply examples/java-api/src/main/java/org/locationtech/geowave/examples/binning/GeohashBinningExample.java]: I added a bunch of comments, let me know if its better, or it should be explained more thoroughly.

File: examples/java-api/src/main/java/org/locationtech/geowave/examples/binning/GeohashBinningExample.java [hunk 26]
```diff
@@ -0,0 +1,123 @@
...
+        "Results for precision 1-6: %s%n",
+        Arrays.toString(allResults.entrySet().toArray(new Map.Entry[0])));
+
+    System.out.printf(
+        "Results just for precision 6: %s%n",
+        Arrays.toString(
+            allResults.entrySet().stream().filter((e) -> e.getKey().length() == 6).toArray()));
+  }
+
+  private static Map<String, BigDecimal> executeBinningAggregation(int precision) {
```
[the remark this answers is not in the document]
Comment (sinistersnare) [reply examples/java-api/src/main/java/org/locationtech/geowave/examples/binning/GeohashBinningExample.java]: Same as above.

File: extensions/adapters/vector/src/main/java/org/locationtech/geowave/adapter/vector/query/aggregation/VectorMathAggregation.java (hunk 5 above)
Comment (sinistersnare) [reply extensions/adapters/vector/src/main/java/org/locationtech/geowave/adapter/vector/query/aggregation/VectorMathAggregation.java:70]: Talked about above, I moved this code over with some slight changes.

File: test/src/test/java/org/locationtech/geowave/test/basic/GeoWaveBinningAggregationIT.java (hunk 7 above)
Comment (sinistersnare) [reply test/src/test/java/org/locationtech/geowave/test/basic/GeoWaveBinningAggregationIT.java:9]: Should be resolved.

File: core/index/src/test/java/org/locationtech/geowave/core/index/VarintUtilsTest.java (hunk 12 above)
Comment (sinistersnare) [reply core/index/src/test/java/org/locationtech/geowave/core/index/VarintUtilsTest.java:231]: Done!

File: examples/java-api/src/main/java/org/locationtech/geowave/examples/binning/GeohashBinningExample.java [hunk 27]
```diff
@@ -0,0 +1,123 @@
+package org.locationtech.geowave.examples.binning;
```
[the remark this answers is not in the document]
Comment (sinistersnare) [reply examples/java-api/src/main/java/org/locationtech/geowave/examples/binning/GeohashBinningExample.java]: Should be resolved.

File: core/store/src/main/java/org/locationtech/geowave/core/store/query/aggregate/BinningAggregation.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (sinistersnare) [reply core/store/src/main/java/org/locationtech/geowave/core/store/query/aggregate/BinningAggregation.java:76]: So, I think that there are valid reasons to be using a precision of 6 (~1B hashes), while only needing a small fraction of the total geohashes that that precision gives.

For example, If I had data of only 1 small country, but it was fairly fine grained, then using precision 6 may be warranted. I may only need ~1 million geohashes of precision 6 to cover the entire country. If I used sizeHint as it I implemented it, it would preallocate a map of 1B elements, which causes an easy OOM.

Under the current system, you may say 'just say maxBins 1,000,000'. But when I do that, I would only get a precision of `4`, which may not be fine-grained enough for the data.

This doesnt solve the preallocation problem, but I think that using maxBins is actually a poor choice for Geohashes, because its global-scale. Also people using geohashes are used to precision, and not used to an arbitrary 'max bins' calculating. If we attached some sort of BBox and did some calculations to figure out, given the BBox and the max bins, what precision can we use, but thats just getting really complicated for geohashes, which are pretty straightforward. 

Im not a geospatial guy, really, so I could be completely off the mark. My experience with geohashes has been that people understand the precision system fairly well. Sorry this became a mini rant cause I dont think I like the maxBins system that much. Of course I dont mind being told 'f-off we are gonna use maxBins' because again, I dont really know what im talking about.

**Comment by rfecher** [comment]
> I signed the ECA a while ago, how can I get confirmation of that?

I wouldn't worry too much about it at this stage, when you complete the PR make sure you sign all your commits with the same email as the ECA and it should pass the ECA validation.

**Comment by rfecher** [comment]
> @rfecher I noticed when reviewing this that most aggregations simply return `new byte[0]` in their serialization functions. It the expectation that an aggregation that is serialized and then deserialized is equivalent to the original, including the value?

@jdgarrett the "Result" and the "Params" are separate from the function itself and serialized separately.  I actually would expect all Aggregations to be empty byte serializations (byte[0]), because the input to the function goes in params anyways.  I _think_ I had good reason to separate serialization of params and the actual function (there's definitely good reason to separate the result). It seems fairly clean and flexible considering the aggregation function is also responsible for serializing/deserializing result and computing the aggregation, there's no need to inherently tie a potentially heavy input parameter to its serialization. By far the most complicated aggregation at this point in the baseline is distributed rendering so making that make reasonable sense drove some of the design decisions (org.locationtech.geowave.adapter.vector.render.DistributedRenderAggregation).

**Comment by sinistersnare** [comment]
I have addressed most of the comments, still some work to be done.

Me and JD had a talk on some different constructors for the GeohashBinningStrategy. Im not quite sure how I can implement them, and I made comments on that front. 

Aside from that, testing, and docs, this is looking better and better.

File: core/store/src/main/java/org/locationtech/geowave/core/store/query/aggregate/AggregationQueryBuilderImpl.java [hunk 28]
```diff
@@ -31,18 +37,34 @@
...
+    this.options.setTypeNames(typeNames);
     return (A) this;
   }
 
-  protected AggregateTypeQueryOptions<P, R, T> newAggregateTypeQueryOptions() {
+  @Override
+  public <A2 extends AggregationQueryBuilder<P, Map<String, R>, T, A2>> AggregationQueryBuilder<P, Map<String, R>, T, A2> bin(
+      AggregationBinningStrategy<T> binningStrategy) {
+    final byte[] aggBytes = PersistenceUtils.toBinary(this.options.getAggregation());
+    Aggregation<P, Map<String, R>, T> metaAgg =
```
[the remark this answers is not in the document]
Comment (sinistersnare) [reply core/store/src/main/java/org/locationtech/geowave/core/store/query/aggregate/AggregationQueryBuilderImpl.java]: I addressed this by adding a new `buildWithBinningStrategy` function. Thanks! Good idea.

**Comment by sinistersnare** [comment]
Signed my commit, rebased onto master (fixing that last TODO!), now we just wait for CI!

Woo!
